src/main/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessor.java

```diff
@@ -155,42 +271,95 @@ public void inferenceSimilarity(
         final List<String> targetResponseFilters,
         final String modelId,
         final List<String> inputText,
+        final MLAlgoParams mlAlgoParams,
         final int retryTime,
         final ActionListener<List<List<Float>>> listener
     ) {
-        MLInput mlInput = createMLTextInput(targetResponseFilters, inputText);
-        mlClient.predict(modelId, mlInput, ActionListener.wrap(mlOutput -> {
-            final List<List<Float>> vector = buildVectorFromResponse(mlOutput);
-            listener.onResponse(vector);
-        }, e -> {
-            if (RetryUtil.shouldRetry(e, retryTime)) {
-                final int retryTimeAdd = retryTime + 1;
-                retryableInferenceSentencesWithVectorResult(targetResponseFilters, modelId, inputText, retryTimeAdd, listener);
-            } else {
-                listener.onFailure(e);
+
+        Consumer<Boolean> runPrediction = isAsymmetricModel -> {
+            MLInput mlInput = createMLTextInput(targetResponseFilters, inputText, isAsymmetricModel ? mlAlgoParams : null);
+            mlClient.predict(modelId, mlInput, ActionListener.wrap(mlOutput -> {
+                final List<List<Float>> vector = buildVectorFromResponse(mlOutput);
+                listener.onResponse(vector);
+            }, e -> {
+                if (RetryUtil.shouldRetry(e, retryTime)) {
+                    final int retryTimeAdd = retryTime + 1;
+                    retryableInferenceSentencesWithVectorResult(
+                        targetResponseFilters,
+                        modelId,
+                        inputText,
+                        mlAlgoParams,
+                        retryTimeAdd,
+                        listener
+                    );
+                } else {
+                    listener.onFailure(e);
+                }
+            }));
+        };
+
+        checkModelAsymmetryAndThenPredict(modelId, listener::onFailure, runPrediction);
+    }
+
+    /**
+     * Check if the model is asymmetric and then run the prediction. Model asymmetry is a concept
+     * that is specific to TextEmbeddingModelConfig. If the model is not a TextEmbeddingModel, then
+     * this check is not applicable.
+     *
+     * The asymmetry of a model is static for a given model. To avoid repeated checks for the same
+     * model, we cache the model asymmetry status. Non-TextEmbeddingModels are cached as false.
+     *
+     * @param modelId    The model id to check
+     * @param onFailure The action to take if the model cannot be retrieved
+     * @param runPrediction The action to take if the model is successfully retrieved
+     */
+    private void checkModelAsymmetryAndThenPredict(String modelId, Consumer<Exception> onFailure, Consumer<Boolean> runPrediction) {
+        CheckedConsumer<MLModel, ? extends Exception> checkModelAsymmetryListener = model -> {
+            MLModelConfig modelConfig = model.getModelConfig();
+            if (!(modelConfig instanceof TextEmbeddingModelConfig)) {
```

Comment:
Add null check for modelConfig as it could be null and throws NPE

Reply:
I believe the null check is not necessary, since the `instanceof` operator is null-safe. Cf. https://docs.oracle.com/javase/specs/jls/se23/html/jls-15.html#jls-15.20.2